Avoid over-eager clipping of child layers in multicolumn

Ahmad-S792 opened the pull request:
#### 1ff6d9ccac00ba79807de18137203a5ec5b2d6f8
<pre>
Avoid over-eager clipping of child layers in multicolumn
<a href="https://bugs.webkit.org/show_bug.cgi?id=272280">https://bugs.webkit.org/show_bug.cgi?id=272280</a>
<a href="https://rdar.apple.com/126413036">rdar://126413036</a>

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/e500cbd3ba36bf3216087481790f5aece0e778a5">https://chromium.googlesource.com/chromium/src.git/+/e500cbd3ba36bf3216087481790f5aece0e778a5</a>

Self-painting layers (caused by e.g. position:relative) don&apos;t contribute to
visual overflow in their parent layout object; see
RenderBox::addOverflowFromChild(). We cannot use the visual overflow rectangle
set on the flow thread when calculating the bounding box of a fragment
established by a child layer.

Therefore, only clip in the flow thread&apos;s block direction in
overflowRectForFragmentedFlowPortion(), and leave the inline axis alone. I
simplified the implementation, since it&apos;s now way easier to start with an
infinite rectangle, and just limit the dimensions that need it afterwards.

Also got rid of an old check for shouldClipFragmentedFlowContent(), which
must have been residue from the CSS regions implementation.

This also happened to fix some inaccuracies mostly invisible to the naked
eye, when it comes to transform:scale()d text with glyphs that have negative
left bearing or overflow the line box vertically. It looks like we measure and
lay out with the CSS computed font, and then switch to a scaled font when
painting, so that it looks crisp and nice.

Also we used &apos;renderableInfiniteRect&apos; similar to other cases in WebKit.
Since it handles any overflows in float scales properly.

* Source/WebCore/rendering/RenderFragmentContainer.cpp:
(WebCore::RenderFragmentContainer::overflowRectForFragmentedFlowPortion const):

&gt; Local Tests:
* LayoutTests/fast/multicol/scale-transform-text.html: Add Test Case
* LayoutTests/fast/multicol/scale-transform-text-expected.html: Add Test Case Expectation
* LayoutTests/fast/multicol/overflowing-relpos.html: Add Test Case
* LayoutTests/fast/multicol/overflowing-relpos-expected.html: Add Test Case Expectation

&gt; WPT Progressions:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-break/relpos-inline-hit-testing-expected.txt:

&gt; Rebaselines:
* LayoutTests/imported/blink/fast/multicol/unbreakable-block-too-tall-at-column-start.html: WPE &amp; GTK Specific Pixel Failures
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-003-expected.txt:
* LayoutTests/fast/multicol/client-rects.html: GTK specific

Canonical link: <a href="https://commits.webkit.org/302249@main">https://commits.webkit.org/302249@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e87b6cf6cb43da08a59128b047ac750fd2f41362

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38495 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134950 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79235 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9023493c-ca0c-480d-a755-0fc57cfe962f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47952 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55857 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97191 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65090 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/101dba08-7496-462a-af45-0913ce55410a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130626 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38345 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114377 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77673 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cbcd9855-1f37-4820-857a-abde70ddf02e) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37144 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32473 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78308 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137432 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54337 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41888 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105712 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54848 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105363 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27047 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50879 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29318 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51951 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54274 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60450 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53509 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56965 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55267 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderFragmentContainer.cpp:
```diff
@@ -156,26 +157,27 @@ LayoutPoint RenderFragmentContainer::fragmentedFlowPortionLocation() const
 
-    LayoutRect fragmentedFlowOverflow = visualOverflowRectForBox(*m_fragmentedFlow);
-    LayoutRect clipRect;
+    auto renderableInfiniteRect = [] {
+        // Return a infinite-like rect whose values are such that, when converted to float pixel values, they can reasonably represent device pixels.
+        return LayoutRect(LayoutUnit::nearlyMin() / 32, LayoutUnit::nearlyMin() / 32, LayoutUnit::nearlyMax() / 16, LayoutUnit::nearlyMax() / 16);
+    }();
```
any particular reason for not using LayoutRect's infiniteRect?

Ahmad-S792 replied:
It leads to WPT failure - css/css-break/out-of-flow-in-multicolumn-018.htm.

Ahmad-S792 replied:
@alanbaradlay - I noticed that Blink has `infiniteIntRect`, which is different from our infiniteRect. I think either we update our or introduce newer helper. I copied this one from Simon's from here - https://searchfox.org/wubkat/rev/4e9806c2e7f7c29e4a6d87ebeca603c4cc8f080c/Source/WebCore/rendering/RenderLayerCompositor.cpp#3630 to workaround here.

alanbaradlay replied:
Let me debug this.

alanbaradlay replied:
I am going to defer to @smfr on this, but 1, it does not scale well 2, technically we now return a non-infinite rect even when there's no clipping at all (so comparing to LayoutRect::infiniteRect() will return false)

Ahmad-S792 committed:
Avoid over-eager clipping of child layers in multicolumn
https://bugs.webkit.org/show_bug.cgi?id=272280
rdar://126413036

Reviewed by Simon Fraser.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/e500cbd3ba36bf3216087481790f5aece0e778a5

Self-painting layers (caused by e.g. position:relative) don't contribute to
visual overflow in their parent layout object; see
RenderBox::addOverflowFromChild(). We cannot use the visual overflow rectangle
set on the flow thread when calculating the bounding box of a fragment
established by a child layer.

Therefore, only clip in the flow thread's block direction in
overflowRectForFragmentedFlowPortion(), and leave the inline axis alone. I
simplified the implementation, since it's now way easier to start with an
infinite rectangle, and just limit the dimensions that need it afterwards.

Also got rid of an old check for shouldClipFragmentedFlowContent(), which
must have been residue from the CSS regions implementation.

This also happened to fix some inaccuracies mostly invisible to the naked
eye, when it comes to transform:scale()d text with glyphs that have negative
left bearing or overflow the line box vertically. It looks like we measure and
lay out with the CSS computed font, and then switch to a scaled font when
painting, so that it looks crisp and nice.

Also we used 'renderableInfiniteRect' similar to other cases in WebKit.
Since it handles any overflows in float scales properly.

* Source/WebCore/rendering/RenderFragmentContainer.cpp:
(WebCore::RenderFragmentContainer::overflowRectForFragmentedFlowPortion const):

> Local Tests:
* LayoutTests/fast/multicol/scale-transform-text.html: Add Test Case
* LayoutTests/fast/multicol/scale-transform-text-expected.html: Add Test Case Expectation
* LayoutTests/fast/multicol/overflowing-relpos.html: Add Test Case
* LayoutTests/fast/multicol/overflowing-relpos-expected.html: Add Test Case Expectation

> WPT Progressions:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-break/relpos-inline-hit-testing-expected.txt:

> Rebaselines:
* LayoutTests/imported/blink/fast/multicol/unbreakable-block-too-tall-at-column-start.html: WPE & GTK Specific Pixel Failures
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-003-expected.txt:
* LayoutTests/fast/multicol/client-rects.html: GTK specific

Canonical link: https://commits.webkit.org/302249@main
